roles/storage-glusterfs/tasks/provision.yml

```diff
@@ -5,31 +5,28 @@
         and admin_user is defined
         and admin_password is defined
 
-- block:
-  - name: Copy glusterfile provisioner YAML
-    copy:
-      src: glusterfile-provisioner.yml
-      dest: /tmp/glusterfile-provisioner.yml
-
-  - name: Deploy glusterfile provisioner pod
-    command: "kubectl apply -n {{ namespace }} -f /tmp/glusterfile-provisioner.yml"
-  when: external_provisioner
+- set_fact:
+    glusterfs_name_label: "{{ glusterfs_name }}{% if glusterfs_name %}-{% endif %}"
 
 - name: Determine the heketi route
-  command: "oc get route {{ heketi_route_name }} -n {{ namespace }} -o jsonpath='{.spec.host}'"
+  command: "oc get route -n {{ glusterfs_namespace }} -l heketi={{ glusterfs_name_label }}route -o jsonpath='{.spec.host}'"
   register: route_cmd
-  when: cluster == "openshift"
+  when:
+    - cluster == 'openshift'
+    - heketi_url is not defined
 
 - set_fact:
     heketi_url: "{{ route_cmd.stdout }}"
   when: route_cmd is defined
 
 - name: Get heketi admin key
-  command: "oc get secret {{ heketi_secret_name }} -n {{ namespace }} -o jsonpath='{.data.key}'"
+  command: "kubectl get pod -n {{ glusterfs_namespace }} -l heketi={{ glusterfs_name_label }}pod -o jsonpath='{.spec.env[?(@.name=\"HEKETI_ADMIN_KEY\")].value}' | base64 -"
```

Comment:
In my environment (OCP 3.9) this did not work unless the jsonpath expression was defined like so:

{.items[0].spec.containers[0].env[?(@.name==\"HEKETI_ADMIN_KEY\")].value}

So you need first to index the .items array, then index into the containers array.  Also '=' needs to be changed to '=='.

Reply:
@rthallisey The heketi pod is part of a Deployment, which autogenerates a random pod name, hence why I search by label.

@aglitke On it.